File: Typescript/src/core/OrderModule.ts
export enum HardwareType {
    Laptop = "Laptop",
    Monitor = "Monitor",
    Desk = "Desk"
}

class APICaller {
    placeOrder(type: HardwareType, number: number): boolean {
        console.log(`Placing order for ${number} ${type}(s)...`);
        return true; // Simulate successful API call
    }
}

class Email {
    to: string;
    from: string;
    header: string;
    body: string;

    constructor(to: string, from: string, header: string, body: string) {
        this.to = to;
        this.from = from;
        this.header = header;
        this.body = body;
    }
}

class Emailer {
    static sendEmail(email: Email): void {
        console.log(`Sending email to: ${email.to}`);
        console.log(`Subject: ${email.header}`);
        console.log(`Body:\n${email.body}`);
    }
}
class EmailComposer {
    static composeEmail(address: string, orderDetails: string, invoiceDetails: string, type: HardwareType): Email {
            const email = new Email(
                address,
                "Ordermodule@example.com",
                `Invoice ${type}`,
                invoiceDetails
            );
            return email;
    }
}
class PriceCalculator {
    static calculatePrice(type: HardwareType, number: number): Number {
        // Calculate price
        let price = 0;
        switch (type) {
            case HardwareType.Laptop:
                price = 1200 * number;
                break;
            case HardwareType.Monitor:
                price = 250 * number;
                break;
            case HardwareType.Desk:
                price = 550 * number;
                break;
            default:
                throw new Error("Invalid hardware type");
        }
        return price;
    }
}
class NumberValidator {
    static validateNumber(type: HardwareType, number: number): Boolean {
        // Validation
        if (number < 1 || number > 30) {
            throw new Error(`Order ${number} of type ${type} seems incorrect`);
            return false;
        }
        return true;
    }
}
export class OrderModule {
    order(type: HardwareType, number: number): void {

        let valid = NumberValidator.validateNumber(type, number);

        if(valid){
            // Order hardware with API
            const apiCaller = new APICaller();
            const result = apiCaller.placeOrder(type, number);
            if (!result) {
                throw new Error("Order failed");
            }

            let price = PriceCalculator.calculatePrice(type, number);

            // Compose and send email
            const address = "itbusiness@example.com";
            const orderDetails = `${number} of ${type}`;
            const invoiceDetails = `Customer email: ${address}\nDetails: ${orderDetails}\nPrice: ${price}`;
            Emailer.sendEmail(EmailComposer.composeEmail(address,orderDetails,invoiceDetails, type));

            console.log("Order processed");
        }
    }
}

// Example usage
const orderModule = new OrderModule();
orderModule.order(HardwareType.Laptop, 3);
orderModule.order(HardwareType.Monitor, 6);
orderModule.order(HardwareType.Desk, 2);

File: CSharp/OrderModule/OrderModule.Core/OrderModule.cs
using OrderModule.Core.Interfaces;
using OrderModule.Core.Services;

namespace OrderModule.Core;

public class OrderModule
    {
        private readonly IOrderValidator _orderValidator;
        private readonly IAPICaller _apiCaller;
        private readonly IEmailService _emailService;
        private readonly ILogger _logger;

        // Expose the last calculated price for testing purposes.
        public decimal LastCalculatedPrice { get; private set; }

        public OrderModule(IOrderValidator orderValidator,
                           IAPICaller apiCaller,
                           IEmailService emailService,
                           ILogger logger)
        {
            _orderValidator = orderValidator;
            _apiCaller = apiCaller;
            _emailService = emailService;
            _logger = logger;
        }

        public void Order(HardwareType type, int number)
        {
            try
            {
                _logger.LogInfo("Starting order process");

                // Validate the order
                _orderValidator.Validate(type, number);
                _logger.LogInfo("Order validated");

                // Place the order via API
                _logger.LogInfo("Placing order via API");
                var apiResult = _apiCaller.PlaceOrder(type, number);
                if (!apiResult.HasSucceeded)
                {
                    _logger.LogError("API order placement failed");
                    throw new Exception("Order failed");
                }
                _logger.LogInfo("API order placement succeeded");

                // Use the API's price result
                LastCalculatedPrice = apiResult.Price;
                _logger.LogInfo($"Price received: {LastCalculatedPrice}");

                // Compose and send email
                var address = "itbusiness@example.com";
                var orderDetails = $"{number} of {type}";
                var invoiceDetails = $"Customer email: {address}\nDetails: {orderDetails}\nPrice: {LastCalculatedPrice}";
                var email = new Email
                {
                    To = address,
                    From = "Ordermodule@example.com",
                    Header = $"Invoice {type}",
                    Body = invoiceDetails,
                };
                _emailService.SendEmail(email);
                _logger.LogInfo("Order email sent");

                Console.WriteLine("Order processed");
                _logger.LogInfo("Order process completed successfully");
            }
            catch (Exception ex)
            {
                _logger.LogError($"Order process encountered an error: {ex.Message}");
                throw;
            }
        }
    }